implementations/rust/ockam/ockam_transport_tcp/src/portal/inlet_shared_state.rs

```diff
@@ -1,25 +1,131 @@
-    route: Route,
-    // Identifier of the other side
-    // The identifier is always the same for the same route as is obtained from the first local
-    // secure channel on the route. However, we should recheck that identifier hasn't changed
-    // when updating the route.
-    their_identifier: Option<LocalInfoIdentifier>,
-    is_paused: bool,
-    // Starts with 0 and increments each time when inlet updates the route to the outlet
-    // (e.g. when reconnecting), this will allow outlet to figure out what is the most recent
-    // return_route even if messages arrive out-of-order
-    route_index: u32,
+    routes: Arc<SyncRwLock<Vec<(String, InletRouteState)>>>,
+    // To notify the listener that a new route is available
+    notify: Arc<Notify>,
 }
 
 impl InletSharedState {
-    pub fn create(ctx: &Context, route: Route, is_paused: bool) -> Result<Self> {
+    /// Add a new route to the shared state and return true if the route was added successfully.
+    /// If the route already exists, the function will return false.
+    pub fn try_add_route(&self, key: String, route_state: InletRouteState) -> bool {
+        let mut guard = self.routes.write().unwrap();
+        if guard.iter().any(|(k, _)| k == &key) {
+            false
+        } else {
+            guard.push((key, route_state));
+            self.notify.notify_waiters();
+            true
+        }
+    }
+
+    /// Adds a new route to the shared state and return an error if the route already exists.
+    pub fn add_route(&self, key: String, route_state: InletRouteState) -> Result<()> {
+        if !self.try_add_route(key, route_state) {
+            Err(ockam_core::Error::new(
+                Origin::Channel,
+                Kind::AlreadyExists,
+                "Route already exists",
+            ))
+        } else {
+            Ok(())
+        }
+    }
+
+    /// List all route keys
+    pub fn list_all_route_keys(&self) -> Vec<String> {
+        let guard = self.routes.read().unwrap();
+        guard.iter().map(|(k, _)| k.clone()).collect()
+    }
+
+    /// Removes a route given its key
+    pub fn remove_route(&self, route_key: &str) {
+        let mut guard = self.routes.write().unwrap();
+        guard.retain(|(k, _)| k != route_key);
+    }
+
+    /// Choose one random active route from the shared state, if no active routes are found, wait
+    /// until a new route is added or an existing route is unpaused.
+    pub async fn choose_active_route(&self) -> InletRouteState {
+        loop {
+            {
+                let guard = self.routes.read().unwrap();
+                let active_routes: Vec<&InletRouteState> = guard
+                    .iter()
+                    .filter(|(_, route)| !route.is_paused())
+                    .map(|(_, route)| route)
+                    .collect();
+
+                if !active_routes.is_empty() {
+                    break active_routes
+                        .choose(&mut rand::thread_rng())
```

Comment:
Could this `rand::thread_rng()` call have a noticeable performance impact if this function becomes a hot path?

Reply:
It should be called once for each new inlet connection, so it shouldn't have any measurable impact. But yeah, we need to be careful with random usage!